Development practices/Turbo frames.md

```diff
@@ -0,0 +1,195 @@
+Turbo frames are one of the [Turbo](https://turbo.hotwired.dev/) techniques which allow predefined parts of a web page to be updated on request.
+They are independent pieces of a web page that can be appended, prepended, replaced, or removed without a complete page refresh and writing a single line of JavaScript.
+
+With Turbo Frames, you can treat a subset of the page as its own component, where links and form submissions replace only that part.
+
+The way it works is: whenever the link inside a frame is clicked or form inside a frame is submitted, the frame content will automatically be updated after receiving a response.
+
+Example:
+
+If we have a part of the page defined like this:
+
+```Html
+  ...
+  <turbo-frame id='id'>
+    <a href='/users/new'>New user</a>
+  </turbo-frame>
+  ...
+```
+
+Once the link to add a new user is clicked, the response might look like this:
+
+```Html
+  <body>
+    <h1> New user </h1>
+
+    <turbo-frame id='id'>
+      <form action='/users'>
+        <input name='user[name]' type='text' value='Name'>
+        <input type="submit">
+      </form>
+    </turbo-frame>
+  </body>
+```
+
+In this case, the frame from the response will be extracted and it will replace the frame with the same id from where the click originated.
+That means that the form for adding a new user will replace the link that was clicked and the rest of the page will remain intact.
+
+## Installation
+To be able to use the turbo frames in our application, the [turbo-rails](https://github.com/hotwired/turbo-rails) gem needs to be installed.
+
+The gem is automatically configured for applications made with Rails 7+ unless --skip-hotwire was passed to the generator.
+
+It can also be installed manually:
+1. Add the turbo-rails gem to your Gemfile: `gem 'turbo-rails'`
+2. Run `./bin/bundle install`
+3. Run `./bin/rails turbo:install`
```

Comment:
when I ran `bin/rails turbo:install` I got this message:

`You must import @hotwired/turbo-rails in your JavaScript entrypoint file`

I ignored it, and following the rest of the examples didn't work (normal page reloads happened). Please add information on how to install turbo frames into:
1. a new application (I had to add webpacker to Gemfile, install it, and add `import '@hotwired/turbo-rails'` to app/javascripts/packs.
2. an existing application - it would be cool to have all the steps necessary for bringing turbo frames to an app that's architectured like one of our apps (with app/webpack folder)

Also, provide a section which explains the dependencies. I believe `webpacker` is one that is required for this to work.

And lastly, is it possible to use Hotwire and its turbo frames with Stimulus? I think we should investigate this as well and add as new sections to this chapter.

Reply:
I just did a triage of our apps and unfortunately, not a lot of our apps use webpacker at all.

But, I think we should have a section where it's explained how to hook up hotwire turbo frames into apps using webpacker, and those that don't.

The be-all end-all is that `import '@hotwired/turbo-rails'` needs to be in the Javascript entrypoint file, which is usually located in `app/javascript/application.js` or `app/webpack/packs/application.js` for apps using webpacker.